chore/check-javadoc-regressions.py

```diff
@@ -0,0 +1,272 @@
+        "checkstyle:check",
+        "--fail-never",
+        f'-Dcheckstyle.config.location={str(config_path)}',
+    ])
+
+
+def extract_violation_count(checkstyle_output: str) -> Optional[int]:
+    """
+    Tries to extract the violation count from the checkstyle output.
+    Returns None if no count could be found.
+    """
+    match = re.search(r"There are (\d+) errors reported by Checkstyle", checkstyle_output)
+    if match:
+        return int(match.group(1))
+    return None
+
+
+def filter_relevant_lines(lines: List[str]) -> List[str]:
+    """Tries to filter the checkstyle output to only include lines with violations."""
+    return [line for line in lines if re.search(r":\d+:", line) and "ERROR" in line]
+
+
+def strip_line_number(line: str) -> str:
+    return re.sub(r".java:\d+(:\d+)?:", ".java", line)
+
+
+def find_added_violation_lines(reference: 'Counter[str]', other: 'Counter[str]') -> 'Counter[str]':
+    """
+    Tries to find out which violation lines are new in "other" and which
+    were already present in the "reference". These lines are added violations the
+    contributor should fix.
+    """
+    # Lines that are only in the reference aren't critical, as they were
+    # apparently fixed in "other"
+    return other - reference
+
+
+def try_readd_line_numbers(without_numbers: 'Counter[str]', with_numbers: List[str]) -> List[str]:
+    """
+    Tries to re-add the line numbers to a dict of strings.
+    This is not necessarily accurate, as it doesn't know which occurrence
+    in "with_numbers" is the real source. If the error message is unique within
+    a file it will find the exact match though.
+    """
+    output: List[str] = []
+
+    for line in with_numbers:
+        stripped = strip_line_number(line)
+        if stripped in without_numbers:
+            if without_numbers[stripped] == 1:
+                without_numbers.pop(stripped)
+            else:
+                without_numbers[stripped] -= 1
+            output.append(line)
+
+    for line in without_numbers:
+        output.append(line)
+
+    return output
+
+
+def print_regression_lines(reference: str, other: str) -> None:
+    """
+    Receives two checkstyle outputs, one for the reference and one for the current branch and
+    prints all identified regressions.
+    """
+    lines_reference = filter_relevant_lines(reference.splitlines())
+    lines_other = filter_relevant_lines(other.splitlines())
+
+    new_errors = find_added_violation_lines(
+        Counter(map(strip_line_number, lines_reference)),
+        Counter(map(strip_line_number, lines_other))
+    )
+    with_line_numbers = try_readd_line_numbers(new_errors, lines_other)
+
+    print(warn("You likely added the following new errors. "), end=" ")
+    print("Line numbers might be incorrect if there are multiple violations of the same type in the file.")
+    for error in with_line_numbers:
+        print(error.strip())
+
+
+def print_compare_with_branch(target_branch: str) -> None:
```

Comment:
This function is pretty large, could you split it into a few smaller ones?

Reply:
I split it up a bit. It still does three distinct things, but they sadly need the output of the previous stages. I could maybe give that some more thought and try to restructure it completely, but I'd leave it as is if you don't want me to do that.

Reply:
I think the new split is perfectly. If this was production code perhaps a more thorough redesign would be warranted, but this is a support script, and by support script standards it's very well put together IMO.